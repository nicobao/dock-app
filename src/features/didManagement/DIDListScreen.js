import React, {useCallback, useState} from 'react';
import {
  BigButton,
  Box,
  Footer,
  Header,
  IconButton,
  NBox,
  ScreenContainer,
  Theme,
  Typography,
} from '../../design-system';
import {translate} from '../../locales';
import {addTestId} from '../../core/automation-utils';
import PlusCircleWhiteIcon from '../../assets/icons/plus-circle-white.svg';
import {RoundedDIDIcon} from '../../assets/icons';
import {ScrollView} from 'native-base';

import {VStack} from 'native-base';
import PlusCircleIcon from '../../assets/icons/plus-circle.svg';
import DocumentDownloadIcon from '../../assets/icons/document-download.svg';
import {navigate} from '../../core/navigation';
import {Routes} from '../../core/routes';
import {DIDListItem} from './components/DIDListItem';
import {SingleDIDOptionsModal} from './components/SingleDIDOptionsModal';
import {useDIDManagementHandlers} from './didHooks';
import {NewDIDModal} from './components/NewDIDModal';
import {pickJSONFile} from '../../core/storage-utils';

export function DIDListScreen({didList, onDeleteDID, onImportDID}) {
  const [isDIDOptionsModalVisible, setIsDIDOptionsModalVisible] =
    useState(false);
  const [isCreateDIDModalVisible, setIsCreateDIDModalVisible] = useState(false);
  const [selectedDID, setSelectedDID] = useState(null);

  return (
    <ScreenContainer {...addTestId('DIDListScreen')} showTabNavigation>
      <Header>
        <Box
          marginLeft={22}
          marginRight={22}
          flexDirection="row"
          alignItems="center">
          <Box flex={1}>
            <Typography fontFamily="Montserrat" fontSize={24} fontWeight="600">
              {translate('app_navigation.did_management')}
            </Typography>
          </Box>
          <Box row>
            <IconButton
              col
              {...addTestId('DIDListScreen_Add_DID')}
              onPress={() => {
                setIsCreateDIDModalVisible(true);
              }}>
              <PlusCircleWhiteIcon
                style={{
                  color: Theme.colors.headerIconColor,
                }}
              />
            </IconButton>
          </Box>
        </Box>
      </Header>
      <ScrollView marginLeft={3} marginRight={3}>
        {didList.length > 0 ? (
          didList.map((didDocumentResolution, i) => {
            return (
              <DIDListItem
                {...addTestId(`DIDListItem_${i}`)}
                key={didDocumentResolution.id}
                onOptionClicked={() => {
                  setSelectedDID(didDocumentResolution);
                  setIsDIDOptionsModalVisible(true);
                }}
                onShare={() => {
                  navigate(Routes.DID_MANAGEMENT_SHARE_DID, {
                    did: didDocumentResolution.didDocument.id,
                    didName: didDocumentResolution.name,
                  });
                }}
                didDocumentResolution={didDocumentResolution}
              />
            );
          })
        ) : (
          <NBox mt={70}>
            <VStack space={7} alignItems="center">
              <RoundedDIDIcon />
              <Typography mt={3} textAlign="center" variant="didDescription">
                {translate('didManagement.did_definition')}
              </Typography>
            </VStack>
          </NBox>
        )}
      </ScrollView>
      {didList.length === 0 ? (
        <Footer marginBottom={114} marginLeft={3} marginRight={3} flex={1}>
          <BigButton
            {...addTestId('CreateNewDID')}
            onPress={() => {
              navigate(Routes.DID_MANAGEMENT_NEW_DID);
            }}
            icon={
              <PlusCircleIcon
                style={{
                  color: Theme.colors.textHighlighted,
                }}
              />
            }>
            {translate('didManagement.create_new_did')}
          </BigButton>
          <BigButton
            {...addTestId('ImportExistingDIDBtn')}
            onPress={onImportDID}
            icon={
              <DocumentDownloadIcon
                style={{
                  color: Theme.colors.textHighlighted,
                }}
              />
            }>
            {translate('didManagement.import_existing_did')}
          </BigButton>
        </Footer>
      ) : null}
      <SingleDIDOptionsModal
        onDeleteDID={onDeleteDID}
        visible={isDIDOptionsModalVisible && selectedDID}
        didDocumentResolution={selectedDID}
        onClose={() => {
          setIsDIDOptionsModalVisible(false);
          setSelectedDID(null);
        }}
      />
      <NewDIDModal
        onImportDID={onImportDID}
        visible={isCreateDIDModalVisible}
        onClose={() => {
          setIsCreateDIDModalVisible(false);
        }}
      />
    </ScreenContainer>
  );
}
export function DIDListScreenContainer({}) {
  const getFile = useCallback(async () => {
    const encryptedJSONWallet = await pickJSONFile();

    if (!encryptedJSONWallet) {
      return;
    }

    navigate(Routes.DID_MANAGEMENT_IMPORT_DID, {
      encryptedJSONWallet,
    });
  }, []);

  const {onDeleteDID, didList} = useDIDManagementHandlers();
  return (
    <DIDListScreen
      didList={didList}
      onDeleteDID={onDeleteDID}
      onImportDID={getFile}
    />
  );
}
