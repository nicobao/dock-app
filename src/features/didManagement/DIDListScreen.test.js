import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {DIDListScreen} from './DIDListScreen';

const mockStore = configureMockStore();

describe('DIDListScreen', () => {
  it('should render correctly without list', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <DIDListScreen didList={[]} onDeleteDID={jest.fn()} />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should render correctly with list', () => {
    const initialState = {
      // placeholder for redux store
    };

    const didList = [
      {
        id: '1657822960364',
        type: 'DIDResolutionResponse',
        didDocument: {
          '@context': [
            'https://www.w3.org/ns/did/v1',
            'https://ns.did.ai/transmute/v1',
            {
              '@base':
                'did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
            },
          ],
          id: 'did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
          verificationMethod: [
            {
              id: '#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
              type: 'JsonWebKey2020',
              controller:
                'did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
              publicKeyJwk: {
                crv: 'Ed25519',
                x: 'vGur-MEOrN6GDLf4TBGHDYAERxkmWOjTbztvG3xP0I8',
                kty: 'OKP',
              },
            },
            {
              id: '#z6LScrLMVd9jvbphPeQkGffSeB99EWSYqAnMg8rGiHCgz5ha',
              type: 'JsonWebKey2020',
              controller:
                'did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
              publicKeyJwk: {
                kty: 'OKP',
                crv: 'X25519',
                x: 'EXXinkMxdA4zGmwpOOpbCXt6Ts6CwyXyEKI3jfHkS3k',
              },
            },
          ],
          authentication: ['#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg'],
          assertionMethod: [
            '#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
          ],
          capabilityInvocation: [
            '#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
          ],
          capabilityDelegation: [
            '#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg',
          ],
          keyAgreement: ['#z6LScrLMVd9jvbphPeQkGffSeB99EWSYqAnMg8rGiHCgz5ha'],
        },
        correlation: [],
      },
    ];
    const wrapper = shallow(
      <DIDListScreen didList={didList} onDeleteDID={jest.fn()} />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
