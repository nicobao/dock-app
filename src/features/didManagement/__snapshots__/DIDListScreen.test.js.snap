// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DIDListScreen should render correctly with list 1`] = `
<RNCSafeAreaView
  accessibilityLabel="DIDListScreen"
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="DIDListScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={22}
          marginRight={22}
        >
          <Box
            flex={1}
          >
            <Styled(Component)
              fontFamily="Montserrat"
              fontSize={24}
              fontWeight="600"
            >
              DIDs
            </Styled(Component)>
          </Box>
          <Box
            row={true}
          >
            <IconButton
              accessibilityLabel="DIDListScreen_Add_DID"
              col={true}
              onPress={[Function]}
              testID="DIDListScreen_Add_DID"
            >
              <SvgFile />
            </IconButton>
          </Box>
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={3}
        marginRight={3}
      >
        <DIDListItem
          accessibilityLabel="DIDListItem_0"
          didDocumentResolution={
            Object {
              "correlation": Array [],
              "didDocument": Object {
                "@context": Array [
                  "https://www.w3.org/ns/did/v1",
                  "https://ns.did.ai/transmute/v1",
                  Object {
                    "@base": "did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                  },
                ],
                "assertionMethod": Array [
                  "#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                ],
                "authentication": Array [
                  "#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                ],
                "capabilityDelegation": Array [
                  "#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                ],
                "capabilityInvocation": Array [
                  "#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                ],
                "id": "did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                "keyAgreement": Array [
                  "#z6LScrLMVd9jvbphPeQkGffSeB99EWSYqAnMg8rGiHCgz5ha",
                ],
                "verificationMethod": Array [
                  Object {
                    "controller": "did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                    "id": "#z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                    "publicKeyJwk": Object {
                      "crv": "Ed25519",
                      "kty": "OKP",
                      "x": "vGur-MEOrN6GDLf4TBGHDYAERxkmWOjTbztvG3xP0I8",
                    },
                    "type": "JsonWebKey2020",
                  },
                  Object {
                    "controller": "did:key:z6Mks8mvCnVx4HQcoq7ZwvpTbMnoRGudHSiEpXhMf6VW8XMg",
                    "id": "#z6LScrLMVd9jvbphPeQkGffSeB99EWSYqAnMg8rGiHCgz5ha",
                    "publicKeyJwk": Object {
                      "crv": "X25519",
                      "kty": "OKP",
                      "x": "EXXinkMxdA4zGmwpOOpbCXt6Ts6CwyXyEKI3jfHkS3k",
                    },
                    "type": "JsonWebKey2020",
                  },
                ],
              },
              "id": "1657822960364",
              "type": "DIDResolutionResponse",
            }
          }
          key="1657822960364"
          onOptionClicked={[Function]}
          onShare={[Function]}
          testID="DIDListItem_0"
        />
      </ForwardRef>
      <SingleDIDOptionsModal
        didDocumentResolution={null}
        onClose={[Function]}
        onDeleteDID={[MockFunction]}
        visible={false}
      />
      <NewDIDModal
        onClose={[Function]}
        visible={false}
      />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;

exports[`DIDListScreen should render correctly without list 1`] = `
<RNCSafeAreaView
  accessibilityLabel="DIDListScreen"
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="DIDListScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={22}
          marginRight={22}
        >
          <Box
            flex={1}
          >
            <Styled(Component)
              fontFamily="Montserrat"
              fontSize={24}
              fontWeight="600"
            >
              DIDs
            </Styled(Component)>
          </Box>
          <Box
            row={true}
          >
            <IconButton
              accessibilityLabel="DIDListScreen_Add_DID"
              col={true}
              onPress={[Function]}
              testID="DIDListScreen_Add_DID"
            >
              <SvgFile />
            </IconButton>
          </Box>
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={3}
        marginRight={3}
      >
        <Memo(ForwardRef(Box))
          mt={70}
        >
          <Memo(ForwardRef(VStack))
            alignItems="center"
            space={7}
          >
            <SvgFile />
            <Styled(Component)
              mt={3}
              textAlign="center"
              variant="didDescription"
            >
              Decentralized Identifiers (DIDs) are a new type of identifier that enables a decentralized digital identity and allows you to accept verifiable credentials.
            </Styled(Component)>
          </Memo(ForwardRef(VStack))>
        </Memo(ForwardRef(Box))>
      </ForwardRef>
      <Styled(Component)
        flex={1}
        marginBottom={114}
        marginLeft={3}
        marginRight={3}
      >
        <BigButton
          accessibilityLabel="CreateNewDID"
          icon={<SvgFile />}
          onPress={[Function]}
          testID="CreateNewDID"
        >
          Create New DID
        </BigButton>
        <BigButton
          accessibilityLabel="ImportExistingDIDBtn"
          icon={<SvgFile />}
          testID="ImportExistingDIDBtn"
        >
          Import existing DID
        </BigButton>
      </Styled(Component)>
      <SingleDIDOptionsModal
        didDocumentResolution={null}
        onClose={[Function]}
        onDeleteDID={[MockFunction]}
        visible={false}
      />
      <NewDIDModal
        onClose={[Function]}
        visible={false}
      />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;
